src/Infrastructure/Persistence/Startup.cs

```diff
@@ -19,26 +20,18 @@ internal static class Startup
 
     internal static IServiceCollection AddPersistence(this IServiceCollection services, IConfiguration config)
     {
-        // TODO: there must be a cleaner way to do IOptions validation...
-        var databaseSettings = config.GetSection(nameof(DatabaseSettings)).Get<DatabaseSettings>();
-        string? rootConnectionString = databaseSettings.ConnectionString;
-        if (string.IsNullOrEmpty(rootConnectionString))
-        {
-            throw new InvalidOperationException("DB ConnectionString is not configured.");
-        }
-
-        string? dbProvider = databaseSettings.DBProvider;
-        if (string.IsNullOrEmpty(dbProvider))
-        {
-            throw new InvalidOperationException("DB Provider is not configured.");
-        }
-
-        _logger.Information($"Current DB Provider : {dbProvider}");
+        services.AddOptions<DatabaseSettings>()
+            .BindConfiguration(nameof(DatabaseSettings))
+            .ValidateDataAnnotations()
+            .ValidateOnStart();
 
         return services
-            .Configure<DatabaseSettings>(config.GetSection(nameof(DatabaseSettings)))
-
-            .AddDbContext<ApplicationDbContext>(m => m.UseDatabase(dbProvider, rootConnectionString))
+            .AddDbContext<ApplicationDbContext>((p, m) =>
+            {
+                var databaseSettings = p.GetRequiredService<IOptions<DatabaseSettings>>().Value;
+                _logger.Information("Current DB Provider : {dbProvider}", databaseSettings.DBProvider);
```

Comment:
isn't this being called each time a dbcontext is created? isn't this gonna log this multiple times then?

Reply:
You are correct, I overlooked this when moving the log statement.